### PR TITLE
[ValueTracking] Consistently propagate `DemandedElts` in ValueTracking functions

### PR DESCRIPTION
**Pull request by goldsteinn**
- **[ValueTracking] Consistently propagate `DemandedElts` is `computeKnownBits`**
- **[ValueTracking] Consistently propagate `DemandedElts` is `isKnownNonZero`**
- **[ValueTracking] Consistently propagate `DemandedElts` is `ComputeNumSignBits`**
- **[ValueTracking] Consistently propagate `DemandedElts` is `computeKnownFPClass`**

### CODE REVIEW

**Comment by goldsteinn** [comment]
This is notably test free and not an NFC. I didn't see any test changes from these and *hope* it won't be necessary to add the 100 odd tests for all the new cases.

My thinking is these cases seem to have just been oversights during original implementation, and if the original commit had had `DemandedElt` propagation, it probably would have gotten on with normal tests.

But LMK.

**Comment by arsenm** [comment]
I don't think there were demanded elts arguments in the original implementations. Should only require an extractelement with a constant index on any test?

**Comment by goldsteinn** [comment]
> I don't think there were demanded elts arguments in the original implementations. Should only require an extractelement with a constant index on any test?

Correct, if there are any cases you think need the coverage I can add (all if necessary).

**Comment by nikic** [comment]
Could you please check whether this has any compile-time impact? (I think the impact might be positive because we don't have to construct a dummy DemandedElts all the time?)

> > I don't think there were demanded elts arguments in the original implementations. Should only require an extractelement with a constant index on any test?
> 
> Correct, if there are any cases you think need the coverage I can add (all if necessary).

Maybe add a test for the vector reverse case? That one is the only one doing something non-trivial...

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 1]
```diff
@@ -2772,8 +2784,11 @@ static bool isKnownNonZeroFromOperator(const Operator *I,
     //    This all implies the 2 i16 elements are non-zero.
     Type *FromTy = I->getOperand(0)->getType();
     if ((FromTy->isIntOrIntVectorTy() || FromTy->isPtrOrPtrVectorTy()) &&
-        (BitWidth % getBitWidth(FromTy->getScalarType(), Q.DL)) == 0)
+        (BitWidth % getBitWidth(FromTy->getScalarType(), Q.DL)) == 0) {
+      if (match(I, m_ElementWiseBitCast(m_Value())))
+        return isKnownNonZero(I->getOperand(0), DemandedElts, Q, Depth);
```
Comment (nikic) [inline llvm/lib/Analysis/ValueTracking.cpp:2789]: I don't think this case is possible? You'd need an element-wise bitcast involving only pointer or integer types. This means it's a pointer-to-pointer bitcast or an integer-to-integer bitcast, which means it must be a trivial bitcast, that we shouldn't bother handling specially.

**Review by nikic — COMMENTED** [review]
Code changes look fine apart from one place.

**Comment by goldsteinn** [comment]
> Maybe add a test for the vector reverse case? That one is the only one doing something non-trivial...

Tests added

**Review by nikic — APPROVED** [review]
LGTM

**Comment by nikic** [comment]
Compile-time is indeed mildly positive: http://llvm-compile-time-tracker.com/compare.php?from=769952d72ff383e0e5fda46802fb8717712784d3&to=0df283c8117daebd019927aff76f78f43cfb5569&stat=instructions%3Au

**Comment by goldsteinn** [comment]
> Could you please check whether this has any compile-time impact? (I think the impact might be positive because we don't have to construct a dummy DemandedElts all the time?)
> 
Seems to be a minor improvement: https://llvm-compile-time-tracker.com/compare.php?from=769952d72ff383e0e5fda46802fb8717712784d3&to=0df283c8117daebd019927aff76f78f43cfb5569&stat=instructions%3Au

**Review by dtcxzyw — APPROVED** [review]
Thank you!

**Pull request closed without merging**